timeutil/schedule.go

```diff
@@ -215,3 +395,341 @@ func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
 
 	return schedule, nil
 }
+
+// ParseScheduleV2 parses a schedule in V2 format. The format is loosely
+// described as:
+//
+//     eventlist = eventset *( ".." eventset )
+//     eventset = wdaylist / timelist / wdaylist "," timelist
+//
+//     wdaylist = wdayset *( "," wdayset )
+//     wdayset = wday / wdayspan
+//     wday =  ( "sun" / "mon" / "tue" / "wed" / "thu" / "fri" / "sat" ) [ DIGIT ]
+//     wdayspan = wday "-" wday
+//
+//     timelist = timeset *( "," timeset )
+//     timeset = time / timespan
+//     time = 2DIGIT ":" 2DIGIT
+//     timespan = time ( "-" / "~" ) time [ "/" ( time / count ) ]
+//     count = 1*DIGIT
+//
+// Special tokens:
+//   weekend	-> sat-sun
+//   day	-> 0:00-24:00
+//
+// Examples:
+// mon,10:00..fri,15:00 (Monday at 10:00, Friday at 15:00)
+// mon,fri,10:00,15:00 (Monday at 10:00 and 15:00, Friday at 10:00 and 15:00)
+// mon-wed,fri,9:00-11:00/2 (Monday to Wednesday and on Friday, twice between
+//                           9:00 and 11:00)
+// mon,9:00~11:00..wed,22:00~23:00 (Monday, sometime between 9:00 and 11:00, and
+//                                  on Wednesday, sometime between 22:00 and 23:00)
+// mon,wed  (Monday and on Wednesday)
+// mon..wed (same as above)
+//
+// Returns a slice of schedules or an error if parsing failed
+func ParseScheduleV2(scheduleSpec string) ([]*Schedule, error) {
+	var schedule []*Schedule
+
+	for _, s := range strings.Split(scheduleSpec, "..") {
+		// cut the schedule in event sets
+		//     eventlist = eventset *( ".." eventset )
+		sched, err := parseEventSetScheduleV2(s)
+		if err != nil {
+			return nil, err
+		}
+		schedule = append(schedule, sched...)
+	}
+	return schedule, nil
+}
+
+// makeWeekdayEvent generates a per day (span) schedule
+func makeWeekdayEvent(start string, end string) (*Schedule, error) {
+	startDay, startMonthWeekDay, _ := parseWeekdaySpec(start)
+	endDay, endMonthWeekDay, _ := parseWeekdaySpec(end)
+
+	if endMonthWeekDay < startMonthWeekDay {
+		return nil, fmt.Errorf("unsupported schedule")
+	}
+
+	sched := &Schedule{
+		Weekday: startDay,
+		// WeekdayEnd:      endDay,
+		MonthWeekday: startMonthWeekDay,
+		// MonthWeekdayEnd: endMonthWeekDay,
+	}
+	if startDay != endDay {
+		sched.WeekdayEnd = endDay
+	}
+	if endMonthWeekDay != startMonthWeekDay {
+		sched.MonthWeekdayEnd = endMonthWeekDay
+	}
+
+	return sched, nil
+}
+
+// parseSpan splits a span string `<start>-<end>` and returns the start and end
+// pieces. If string is not a span then the whole string is returned
+func parseSpan(spec string) (string, string) {
```

Comment:
Doesn't feel worth having a function for this. Any users of this can just call `strings.SplitN(s, "-", 2)` and have the two pieces at hand trivially. For times, we can call `strings.IndexAny(s, "-~")`. Both of these are one-liners and would sit comfortably inside their context (parseTimeSpan, for example).